master/internal/resourcemanagers/setup.go

```diff
@@ -0,0 +1,74 @@
+package resourcemanagers
+
+import (
+	"crypto/tls"
+
+	"github.com/labstack/echo"
+	"github.com/sirupsen/logrus"
+
+	"github.com/determined-ai/determined/master/internal/agent"
+	"github.com/determined-ai/determined/master/internal/kubernetes"
+
+	"github.com/determined-ai/determined/master/pkg/actor"
+)
+
+// Setup setups the actor and endpoints for resource managers.
+func Setup(
+	system *actor.System,
+	echo *echo.Echo,
+	rmConfig *ResourceManagerConfig,
+	poolsConfig *ResourcePoolsConfig,
+	cert *tls.Certificate,
+) *actor.Ref {
+	var ref *actor.Ref
+	switch {
+	case rmConfig.AgentRM != nil:
+		ref = setupAgentResourceManager(system, echo, rmConfig.AgentRM, poolsConfig, cert)
+	case rmConfig.KubernetesRM != nil:
+		ref = setupKubernetesResourceManager(system, echo, rmConfig.KubernetesRM)
+	default:
+		panic("no expected resource manager config is defined")
+	}
+
+	rm, ok := system.ActorOf(actor.Addr("resourceManagers"), &ResourceManagers{ref: ref})
+	if !ok {
+		panic("cannot create resource managers")
+	}
+	return rm
+}
+
+func setupAgentResourceManager(
+	system *actor.System,
+	echo *echo.Echo,
+	rmConfig *AgentResourceManagerConfig,
+	poolsConfig *ResourcePoolsConfig,
+	cert *tls.Certificate,
+) *actor.Ref {
+	ref, _ := system.ActorOf(
+		actor.Addr("agentRM"),
+		newAgentResourceManager(rmConfig, poolsConfig, cert),
+	)
+	system.Ask(ref, actor.Ping{}).Get()
```

Comment:
question: why do we need this ping?

Reply:
Because when an agent connects to the master, the master looks for the resource pool actor, which is created during starting up the agent resource manager actor. So the agent resource manager actor needs to finish prestart phase before the master listens to the `agents` endpoint. When the `Ping` message is responded by the actor system, the actor should already finish processing `PreStart`.